packages/protocol/contracts-0.8/common/EpochManagerInitializer.sol

```diff
@@ -0,0 +1,62 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity >=0.8.7 <0.8.20;
+
+import "../common/UsingRegistry.sol";
+import "../common/UsingPrecompiles.sol";
+
+import "../../contracts/common/Initializable.sol";
+import "../../contracts/common/interfaces/ICeloVersionedContract.sol";
+import "../../contracts/governance/interfaces/IEpochRewards.sol";
+
+contract EpochManagerInitializer is Initializable, UsingPrecompiles, UsingRegistry {
+  /**
+   * @notice Sets initialized == true on implementation contracts
+   * @param test Set to true to skip implementation initialization
+   */
+  constructor(bool test) public Initializable(test) {}
+
+  /**
+   * @notice Used in place of the constructor to allow the contract to be upgradable via proxy.
+   * @param registryAddress The address of the registry core smart contract.
+   */
+  function initialize(address registryAddress) external initializer {
+    _transferOwnership(msg.sender);
+    setRegistry(registryAddress);
+  }
+
+  /**
+   * @notice initializes the epochManager contract during L2 transition.
+   */
+  function initEpochManager() external onlyOwner {
```

Comment:
should be split into 2 functions:

1. `captureEpochAndValidators()` which is permissionless and can be called while `isOnL2` is false. This function captures: validators and epoch number
2. `initEpochManager()` it's permissionless and can only be called when `isOnL2` is true. This function uses captured data and calls `EpochManager.initilizae()`

Then

Function `initEpochManager()` can only be called once, since EpochManager won't allow 2 calls to `initialize()`

Reply:
Created an [issue](https://github.com/celo-org/celo-blockchain-planning/issues/590) tracking changes to the epochManagerInitializer